core/sql/src/main/java/org/trafodion/sql/HTableClient.java

```diff
@@ -421,29 +839,140 @@ else if (versions > 0)
+		    	  }
+		      }	    	
+		    }
+		    if (columnsToRemove != null)
+		    { //we are almost done checking if Exclude version of SingleColumnnValueFilter can be used. Th elast check s about to know if there is a IS_NULL_NULL
+		      //operation that cannot be using the Exclude method, as it is transformed in a filterList with OR, therefore we cannot guaranty that the SingleColumnValueExcludeFilter
+		      //performing the exclusion will be reached.
+		    	boolean is_null_nullFound = false;
+		    	for (Object o:compareOpList ){
+		    		if (new String((byte[])o).equals("IS_NULL_NULL")){
+		    			is_null_nullFound = true;
+		    			break;
+		    		}		    			
+		    	}
+		    	if (is_null_nullFound){
+		    		columnsToRemove = null; // disable Exclude method version of SingleColumnnValueFilter
+		    	}else
+		    		narrowDownResultColumns = false; // we will use the Exclude version of SingleColumnnValueFilter, so bypass the Family/QualifierFilter method
+		    }
+		    Filter f =constructV2Filter(colNamesToFilter,compareOpList,colValuesToCompare, columnsToRemove);
+		    if (f==null) return false; // error logging done inside constructV2Filter
+		    list.addFilter(f);
+		}//end V2
+		else{// deal with V1
+		    list = new FilterList(FilterList.Operator.MUST_PASS_ALL);
+		    
+		    for (int i = 0; i < colNamesToFilter.length; i++) {
+		      byte[] colName = (byte[])colNamesToFilter[i];
+		      byte[] coByte = (byte[])compareOpList[i];
+		      byte[] colVal = (byte[])colValuesToCompare[i];
+	
+		      if ((coByte == null) || (colVal == null)) {
+		        return false;
+		      }
+		      String coStr = new String(coByte);
+		      CompareOp co = CompareOp.valueOf(coStr);
+	
+		      SingleColumnValueFilter filter1 = 
+		          new SingleColumnValueFilter(getFamily(colName), getName(colName), 
+		              co, colVal);
+		      list.addFilter(filter1);
+		    }			
+		}//end V1
+	    // if we added a column for predicate eval, we need to filter down result columns
+	    FilterList resultColumnsOnlyFilter = null;
+	    if (narrowDownResultColumns){  	    	
+  		    HashMap<String,ArrayList<byte[]>> hm = new HashMap<String,ArrayList<byte[]>>(3);//use to deal with multiple family table
+  		    // initialize hm with list of columns requested for output
+  		    	for (int i=0; i<columns.length; i++){ // if we are here we know columns is not null
+  		    		if (hm.containsKey(new String(getFamily((byte[])columns[i])))){
+  		    			hm.get(new String(getFamily((byte[])columns[i]))).add((byte[])columns[i]);
+  		    		}else{
+  		    			ArrayList<byte[]> al = new ArrayList<byte[]>();
+  		    			al.add((byte[])columns[i]);
+  		    			hm.put(new String(getFamily((byte[])columns[i])), al);
+  		    		}	    			
+  		    	}
+  		    	
+  	    	if (hm.size()==1){//only one column family
+  	    		resultColumnsOnlyFilter = new FilterList(FilterList.Operator.MUST_PASS_ALL);
```

Comment:
in this explicit example, the code is optimized not to use qualifier and family filter, instead it used the SingleColumnValueExcludeFilter, therefore not adding the overhead. Now if instead you do select a,b from t where B>10 and B<20 and c<5, then the optimization is not applicable and yes, I use the qualifier/family filter to pull in a and b only. Note that if a is nullable, and b is nullable, then I will also add one column from the key, avoiding _SALT_ if possible to make sure "absent" null are correctly accounted for. The column to retrieved is computed in HbaseAccess::computeRetrievedCols() that you may not see since it did not change. The trick is that by moving the predicate pushdown computing before invoking computeRetrievedCols in V2, the executorPred() (remaining predicate that cannot be pushed down if any) is already containing nothing if predates will be pushed down. So column computation will not include columns that are needed for predicate evaluation at hbase level. Now for the operator evaluation think you are right. Will have to experiment with it and see where is the point of diminishing return. One thing to note is that this is the last item on the filter list on purpose, so this is not evaluated on filtered out data. Had same concern and tried to put all mitigation possible... This whole thing about limiting data is intuitely attractive, but may be counter productive (CPU intense vs memory/network IO friendly). But given HBase implemented these filter out of the box, I assume they must have seen a benefit...